universal-application-tool-0.0.1/app/controllers/applicant/RedirectController.java

```diff
@@ -0,0 +1,88 @@
+import org.pac4j.play.java.Secure;
+import play.i18n.MessagesApi;
+import play.mvc.Http;
+import play.mvc.Result;
+import repository.ProgramRepository;
+import views.applicant.ApplicantUpsellCreateAccountView;
+
+public class RedirectController extends CiviFormController {
+  private final ProfileUtils profileUtils;
+  private final ProgramRepository programRepository;
+  private final ApplicantUpsellCreateAccountView upsellView;
+  private final MessagesApi messagesApi;
+
+  @Inject
+  public RedirectController(
+      ProfileUtils profileUtils,
+      ProgramRepository programRepository,
+      ApplicantUpsellCreateAccountView upsellView,
+      MessagesApi messagesApi) {
+    this.profileUtils = Preconditions.checkNotNull(profileUtils);
+    this.programRepository = Preconditions.checkNotNull(programRepository);
+    this.upsellView = Preconditions.checkNotNull(upsellView);
+    this.messagesApi = Preconditions.checkNotNull(messagesApi);
+  }
+
+  @Secure
+  public CompletableFuture<Result> programByName(Http.Request request, String programName) {
+    Optional<UatProfile> profile = profileUtils.currentUserProfile(request);
+    if (profile.isEmpty()) {
+      return CompletableFuture.completedFuture(
+          redirect(routes.CallbackController.callback("GuestClient")));
+    }
+    CompletableFuture<Applicant> applicant = profile.get().getApplicant();
+    CompletableFuture<Program> program = programRepository.getForSlug(programName);
+    return CompletableFuture.allOf(applicant, program)
+        .thenApply(
+            empty -> {
+              if (applicant.isCompletedExceptionally()) {
+                return notFound();
+              } else if (program.isCompletedExceptionally()) {
+                return notFound();
+              }
+              return redirect(
+                  controllers.applicant.routes.ApplicantProgramsController.edit(
+                      applicant.join().id, program.join().id));
+            });
+  }
+
+  @Secure
+  public CompletableFuture<Result> considerSignIn(Http.Request request, String redirectTo) {
+    Optional<UatProfile> profile = profileUtils.currentUserProfile(request);
+    if (profile.isEmpty()) {
+      // should definitely never happen.
+      return CompletableFuture.completedFuture(
+          badRequest("You are not signed in - you cannot perform this action."));
+    }
+    return profile
+        .get()
+        .getAccount()
+        .thenApplyAsync(
+            account -> {
+              // Don't show this page to TIs, or anyone with an email address already.
+              if (!Strings.isNullOrEmpty(account.getEmailAddress())) {
+                return redirect(redirectTo);
+              } else if (account.getMemberOfGroup().isPresent()) {
+                return redirect(redirectTo);
+              }
+
+              // To make this happen we really need to be pretty sure we are able to redirect
+              // correctly
+              // after a login!
```

Comment:
Nit (blocking): mind fixing the line break?

Reply:
That was a note to self, even, so removed.